schemas/authoring-guide.md

```diff
@@ -0,0 +1,829 @@
+
+### Limitations of Union Discrimination
+
+Authoring Unions in IPLD Schemas help expose some of the limitations of quickly validating data that is allowed to vary. If we extend our example and introduce another acceptable form of `"payload"` we can see how this ability to quickly discriminate breaks down and introduces the need to do child-contents checking to discriminate:
+
+```json
+{
+  "msg": "Ping",
+  "payload": {
+    "ts": 1572935564043,
+    "nonce": "424f524b"
+  }
+}
+```
+
+We've introduced a new message type but lost the ability to discriminate based in kind as our new type is also a Map. A Schema that accommodates for this additional payload type is possible but forces the burden of discrimination and onto the consumer of the data as well as some additional validation burden:
+
+```ipldsch
+type Message struct {
+  msg String
+  payload Payload
+}
+
+type Payload union {
+	| Error string
+	| ProgressOrPing map
+} representation kinded
+
+type Error string
+
+type ProgressOrPing struct {
+  percent optional Float
+  last optional String
+  ts optional Int
+  nonce optional String
+}
+```
+
+Now the user of such a Schema must do their own field inspection to determine whether a `ProgressOrPing` is a progress message or a ping. Additionally, the burden of ensuring that both `percent` and `last` are present _or_ `ts` and `nonce` are present is left to the user, the Schema layer can't help here. The trade-off present in this scenario regards validation of a node by inspection of its child nodes. Type type of data is common in the real world but IPLD Schemas encourage better data shape design to allow for fast validation through clear discrimination where such variance exists.
+
+### Alternative Discrimination Strategies
+
+If we are designing the data layout for our example protocol (rather than consuming something we have no control over the design of), we could choose a alternate strategy that would allow more efficient discrimination. Unions allow for five different representation strategies that allow for different kinds of discrimination.
+
+#### Keyed
+
+If our generic `"payload"` could be replaced with a specific key that discriminates the type of the payload, we could use a `keyed` Union:
+
+```json
+{
+  "msg": "Something bad happened",
+  "error": "ERROR"
+}
+```
+
+```json
+{
+  "msg": "All good",
+  "progress": {
+    "percent": 0.6,
+    "last": "61626378797a"
+  }
+}
+```
+
+```json
+{
+  "msg": "Ping",
+  "ping": {
+    "ts": 1572935564043,
+    "nonce": "424f524b"
+  }
+}
+```
+
+We can now easily handle this data with the following Schema:
+
+```ipldsch
+type Message struct {
+  msg String
+  payload Payload
+}
+
+type Payload union {
+	| Error "error"
+	| Progress "progress"
+  | Ping "ping"
```

Comment:
it's the tab vs space thing, going with spaces in here, it's my doc so my rules, tabs be damned!

Reply:
"Tabs for indentation, spaces for alignment" is a rule I favor because it works consistently in all environments and preferences.

(Though docs and examples occasionally get a special pass just because screen realestate premiums in docs can be unusually high...)

Reply:
Tere's my rule: tabs where the context demands and where my editor makes it easy, spaces otherwise.
Almost everything I do is tabless so my editors all do 2-spaces when I press `tab`. In these docs where I need a tab I have to go copy it from somewhere and paste it in! Go and Makefiles are the special cases but thankfully my editors can deal with that.